cql3/expr/term_expr.cc

```diff
@@ -453,49 +478,55 @@ tuple_constructor_test_assignment(const tuple_constructor& tc, database& db, con
 }
 
 static
-shared_ptr<term>
+expression
 tuple_constructor_prepare_nontuple(const tuple_constructor& tc, database& db, const sstring& keyspace, lw_shared_ptr<column_specification> receiver) {
     tuple_constructor_validate_assignable_to(tc, db, keyspace, *receiver);
-    std::vector<shared_ptr<term>> values;
+    std::vector<expression> values;
     bool all_terminal = true;
     for (size_t i = 0; i < tc.elements.size(); ++i) {
-        auto&& value = prepare_term(tc.elements[i], db, keyspace, component_spec_of(*receiver, i));
-        if (dynamic_pointer_cast<non_terminal>(value)) {
+        expression value = prepare_term(tc.elements[i], db, keyspace, component_spec_of(*receiver, i));
+        if (!is<constant>(value)) {
             all_terminal = false;
         }
         values.push_back(std::move(value));
     }
-    tuples::delayed_value value(static_pointer_cast<const tuple_type_impl>(receiver->type), values);
+    tuple_constructor value {
+        .elements  = std::move(values),
+        .type = static_pointer_cast<const tuple_type_impl>(receiver->type)
+    };
     if (all_terminal) {
-        return value.bind(query_options::DEFAULT);
+        return evaluate(value, query_options::DEFAULT);
     } else {
-        return make_shared<tuples::delayed_value>(std::move(value));
+        return value;
     }
 }
 
 static
-shared_ptr<term>
+expression
 tuple_constructor_prepare_tuple(const tuple_constructor& tc, database& db, const sstring& keyspace, const std::vector<lw_shared_ptr<column_specification>>& receivers) {
     if (tc.elements.size() != receivers.size()) {
         throw exceptions::invalid_request_exception(format("Expected {:d} elements in value tuple, but got {:d}: {}", receivers.size(), tc.elements.size(), tc));
     }
 
-    std::vector<shared_ptr<term>> values;
+    std::vector<expression> values;
     std::vector<data_type> types;
     bool all_terminal = true;
     for (size_t i = 0; i < tc.elements.size(); ++i) {
-        auto&& t = prepare_term(tc.elements[i], db, keyspace, receivers[i]);
-        if (dynamic_pointer_cast<non_terminal>(t)) {
+        expression elem = prepare_term(tc.elements[i], db, keyspace, receivers[i]);
+        if (!is<constant>(elem)) {
             all_terminal = false;
         }
```

Comment:
We can consider moving this all_terminal stuff to a separate pass where it can be dealt with in a systematic manner.